cpp/src/arrow/builder.cc

```diff
@@ -81,35 +81,29 @@ Status ArrayBuilder::AppendToBitmap(const uint8_t* valid_bytes, int64_t length)
   return Status::OK();
 }
 
-Status ArrayBuilder::Init(int64_t capacity) {
-  int64_t to_alloc = BitUtil::BytesForBits(capacity);
-  null_bitmap_ = std::make_shared<PoolBuffer>(pool_);
-  RETURN_NOT_OK(null_bitmap_->Resize(to_alloc));
-
-  // Buffers might allocate more then necessary to satisfy padding requirements
-  const int64_t byte_capacity = null_bitmap_->capacity();
-  capacity_ = capacity;
-  null_bitmap_data_ = null_bitmap_->mutable_data();
-  memset(null_bitmap_data_, 0, static_cast<size_t>(byte_capacity));
-
-  return Status::OK();
-}
-
-Status ArrayBuilder::Resize(int64_t new_bits) {
+Status ArrayBuilder::Resize(int64_t capacity) {
   if (!null_bitmap_) {
-    return Init(new_bits);
-  }
-  int64_t new_bytes = BitUtil::BytesForBits(new_bits);
-  int64_t old_bytes = null_bitmap_->size();
-  RETURN_NOT_OK(null_bitmap_->Resize(new_bytes));
-  null_bitmap_data_ = null_bitmap_->mutable_data();
-  // The buffer might be overpadded to deal with padding according to the spec
-  const int64_t byte_capacity = null_bitmap_->capacity();
-  capacity_ = new_bits;
-  if (old_bytes < new_bytes) {
-    memset(null_bitmap_data_ + old_bytes, 0,
-           static_cast<size_t>(byte_capacity - old_bytes));
+    int64_t to_alloc = BitUtil::BytesForBits(capacity);
+    null_bitmap_ = std::make_shared<PoolBuffer>(pool_);
+    RETURN_NOT_OK(null_bitmap_->Resize(to_alloc));
+
+    // Buffers might allocate more then necessary to satisfy padding requirements
+    const int64_t byte_capacity = null_bitmap_->capacity();
+    null_bitmap_data_ = null_bitmap_->mutable_data();
+    memset(null_bitmap_data_, 0, static_cast<size_t>(byte_capacity));
+  } else {
+    int64_t new_bytes = BitUtil::BytesForBits(capacity);
+    int64_t old_bytes = null_bitmap_->size();
```

Comment:
I think `->capacity()` makes more sense here, since we are always zeroing up-to allocated capacity. It's doesn't lead to errors, just zeroes the padding twice.

Reply:
Changed